clients/vscode-hlasmplugin/src/hlasmDownloadCommands.ts

```diff
@@ -21,41 +21,103 @@ import { promises as fsp } from "fs";
 import { stripJsonComments } from './tools.common';
 
 export type JobId = string;
-export interface JobDescription {
+export type JobDescription = {
     jobname: string;
     id: JobId;
-    details: string;
-}
-
-function getJobDetailInfo(job: JobDescription): { rc: number; spoolFiles: number } | undefined {
-    const parsed = /^.*RC=(\d+)\s+(\d+) spool file/.exec(job.details);
+} & ({
+    rc: undefined;
+    spoolFiles: undefined;
+} | {
+    rc: number;
+    spoolFiles: string;
+});
+
+function getJobDetailInfo(jobDetailString: string): { rc: number; spoolFiles: string } | { rc: undefined; spoolFiles: undefined } {
+    const parsed = /^.*RC=(\d+)\s+(\d+) spool file/.exec(jobDetailString);
     if (!parsed)
-        return undefined;
+        return { rc: undefined, spoolFiles: undefined };
     else
-        return { rc: +parsed[1], spoolFiles: +parsed[2] };
+        return { rc: +parsed[1], spoolFiles: parsed[2] };
+}
+
+export function ftpJobMapper(fi: FileInfo): JobDescription {
+    const parsedLine = /(\S+)\s+(\S+)\s+(.*)/.exec(fi.name);
+    if (!parsedLine)
+        throw Error("Unable to parse the job list");
+    return { jobname: parsedLine[1], id: parsedLine[2], ...getJobDetailInfo(parsedLine[3]) }
 }
+
+export function extractJobId(s: string) {
+    const jobid = /^.*as ([Jj](?:[Oo][Bb])?\d+)/.exec(s);
+    if (!jobid)
+        throw Error("Unable to extract the job id");
+    return jobid[1];
+}
+
 export interface JobClient {
     submitJcl(jcl: string): Promise<JobId>;
     setListMask(mask: string): Promise<void>;
     list(): Promise<JobDescription[]>;
-    download(target: Writable | string, id: JobId, spoolFile: number): Promise<void>;
+    download(target: Writable | string, job: JobDescription): Promise<void>;
     dispose(): void;
 }
 
-async function basicFtpJobClient(connection: {
+export function zoweJobMapper(spoolObj: string, jobs: {
+    jobid: string;
+    jobname: string;
+    retcode?: string;
+}[]): JobDescription[] {
+    return jobs.map(x => ({
+        jobname: x.jobname,
+        id: x.jobid,
+        ...x.retcode?.startsWith("CC ")
+            ? { rc: +x.retcode.substring(3), spoolFiles: spoolObj }
+            : { rc: undefined, spoolFiles: undefined }
+    }));
+}
+
+async function zoweJobClient(stepDD: string, ci: ZoweConnectionInfo): Promise<JobClient> {
+    const jes = await ensureValidMfZoweClient<any>(ci, ci.zoweExplorerApi.getJesApi);
+    let prefix = '';
+
+    return {
+        async submitJcl(jcl: string): Promise<JobId> { return jes.submitJcl(jcl).then((x: { jobid: string }) => x.jobid); },
+        async setListMask(mask: string): Promise<void> { prefix = mask; },
+        async list(): Promise<JobDescription[]> { return jes.getJobsByParameters({ prefix, status: 'OUTPUT' }).then(zoweJobMapper.bind(undefined, stepDD)); },
+        async download(target: Writable | string, job: JobDescription): Promise<void> {
+            if (typeof target === 'string') throw Error('Unexpected target');
```

Comment:
I don't know much about Writables but to help me understand the program flow why is there a check for `string` in this case?

Reply:
This looks like a remnant of earlier implementations... I'll remove it.